src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -167,25 +164,141 @@ public static unsafe void MatMulX(bool add, AlignedArray mat, AlignedArray src,
+                            }
+                        }
+                        else
+                        {
+                            if (misalignment != 0)
+                            {
+                                // Handle cases where the data is not 256-bit aligned by doing an unaligned read and then
+                                // masking any elements that will be included in the first aligned read
+                                misalignment >>= 2;
+                                misalignment = 8 - misalignment;
+
+                                float* pMatTemp = pMatCurrent;
+                                Vector256<float> x01 = Avx.LoadVector256(pMatTemp);
+                                Vector256<float> x11 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x21 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> x31 = Avx.LoadVector256(pMatTemp += ccol);
+                                Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);
+
+                                Vector256<float> mask = Avx.LoadVector256(((float*)(pLeadingAlignmentMask)) + (misalignment * 8));
+
+                                Vector256<float> tempX01 = Avx.And(x01, mask);
+                                Vector256<float> tempX11 = Avx.And(x11, mask);
+                                Vector256<float> tempX21 = Avx.And(x21, mask);
+                                Vector256<float> tempX31 = Avx.And(x31, mask);
+
+                                Vector256<float> tempVec = Avx.And(vector, mask);
+
+                                res0 = Avx.Multiply(tempX01, tempVec);
+                                res1 = Avx.Multiply(tempX11, tempVec);
+                                res2 = Avx.Multiply(tempX21, tempVec);
+                                res3 = Avx.Multiply(tempX31, tempVec);
+
+                                pMatCurrent += misalignment;
+                                pSrcCurrent += misalignment;
+                                length -= misalignment;
+                            }
+
+                            if (length > 7)
+                            {
+                                remainder = length % 8;
+                                while (pSrcCurrent + 8 <= pSrcEnd)
+                                {
+                                    float* pMatTemp = pMatCurrent;
+                                    Vector256<float> x01 = Avx.LoadVector256(pMatTemp);
+                                    Vector256<float> x11 = Avx.LoadVector256(pMatTemp += ccol);
+                                    Vector256<float> x21 = Avx.LoadVector256(pMatTemp += ccol);
+                                    Vector256<float> x31 = Avx.LoadVector256(pMatTemp += ccol);
+                                    Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);
+
+                                    res0 = Avx.Add(res0, Avx.Multiply(x01, vector));
```

Comment:
same here, we will probably get better throughput by loading vector first and trying to load `x01..x31` as part of the multiply operation